numpy/_core/src/umath/umathmodule.c

```diff
@@ -144,15 +145,75 @@ ufunc_frompyfunc(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds) {
     /* Do a better job someday */
     doc = "dynamic ufunc based on a python function";
 
+    /* Create the ufunc but without any loops by not passing types*/
     self = (PyUFuncObject *)PyUFunc_FromFuncAndDataAndSignatureAndIdentity(
             (PyUFuncGenericFunction *)pyfunc_functions, data,
-            types, /* ntypes */ 1, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
+            NULL, /* ntypes */ 0, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
             str, doc, /* unused */ 0, NULL, identity);
 
     if (self == NULL) {
         PyArray_free(ptr);
         return NULL;
     }
+
+    PyArray_DTypeMeta *op_dtypes[NPY_MAXARGS];
+    for (int arg = 0; arg < nin + nout; arg++) {
+        op_dtypes[arg] = PyArray_DTypeFromTypeNum(types[arg]);
+        /* These DTypes are immortal and adding INCREFs: so borrow it */
+        Py_DECREF(op_dtypes[arg]);
+    }
+
+    char method_name[101];
+    snprintf(method_name, 100, "legacy_ufunc_wrapper_for_%s", str);
+
+    NPY_ARRAYMETHOD_FLAGS flags = NPY_METH_REQUIRES_PYAPI | NPY_METH_NO_FLOATINGPOINT_ERRORS;
+
+    PyArrayMethod_GetReductionInitial *get_reduction_intial = NULL;
+    if (nin == 2 && nout == 1) {
+        npy_bool reorderable = NPY_FALSE;
+        PyObject *identity_obj = PyUFunc_GetDefaultIdentity(
+                self, &reorderable);
+        if (identity_obj == NULL) {
+            PyArray_free(ptr);
+            return NULL;
+        }
+        /*
+         * TODO: For object, "reorderable" is needed(?), because otherwise
+         *       we disable multi-axis reductions `arr.sum(0, 1)`. But for
+         *       `arr = array([["a", "b"], ["c", "d"]], dtype="object")`
+         *       it isn't actually reorderable (order changes result).
+         */
+        if (reorderable) {
+            flags |= NPY_METH_IS_REORDERABLE;
+        }
+        if (identity_obj != Py_None) {
+            get_reduction_intial = &get_initial_from_ufunc;
+        }
+    }
+
+    PyType_Slot slots[4] = {
+        {NPY_METH_get_loop, &get_wrapped_legacy_ufunc_loop},
```

Comment:
Hmmm, would be nice to just pass a new-style loop directly.  That means adding `context` as first and `NpyAuxData *unused` as last entry.

Unfortunately, we do need to store the Python function somewhere, though.  Which would mean either using the auxdata (a bit tedious) or reaching into the internals and setting `static_data` (maybe a bit tedious, because we need to create the `PyArrayMethod` first).

Reply:
Let me see if I am understanding correctly what would be required for each option. If we use NpyAuxData, then we would need to define a new get_strided_loop function that would pass the reference to the python function in the auxdata parameter? Or if we use static_data, we just need to add a _NPY_METH_static_data entry to the slots with the reference and then the NPY_METH_strided_loop could get this via context->method->static_data. So long as the static_data is not considered a do not use internal mechanism, it seems like the simpler approach. Do you agree?

Reply:
Should have scrolled down first :).  Yes, exactly, using the static data is best.
(This is internal right now, but we should make it public probably maybe by copying it to `context->static_data` explicitly for convenience).
(And once it's there, it we may want to add a `free_static_data` as well.)